File: Sem03/SEP3/CSharp/ClientBlazor/Models/Provider.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using GrpcFileGeneration.Models;
using Networking.Provider;
using RiskFirst.Hateoas.Models;

namespace ClientBlazor.Models
{
    public class Provider : User, ILinkContainer
    {
        public Dictionary<string, Link> Links { set; get; }
        
        [Required, MaxLength(50)]
        public string CompanyName { get; set; }
        [Required, Range(10000000,99999999)]
        public int Cvr { get; set; }
        [Required, MaxLength(20)]
        public string PhoneNumber { get; set; }
        [Required, MaxLength(500)]
        public string Description { get; set; }
        
        public bool IsApproved { set; get; }

        [ValidateComplexType]
        public Address Address { get; set; }


        public Provider()
        {
            Links = new Dictionary<string, Link>();
            Address = new Address();
        }

        public Provider(Provider provider)
        {
            Address = provider.Address;
            Cvr = provider.Cvr;
            Description = provider.Description;
            Links = new Dictionary<string, Link>();
            CompanyName = provider.CompanyName;
            IsApproved = provider.IsApproved;
            PhoneNumber = provider.PhoneNumber;
            Email = provider.Email;
            Id = provider.Id;
            Password = provider.Password;
            Token = provider.Token;
            SecurityType = provider.SecurityType;
        }

        public Provider(ProviderMessage message) : base(message.User)
        {
            Links = new Dictionary<string, Link>();
            CompanyName = message.CompanyName;
            Cvr = message.Cvr;
            Description = message.Description;
            PhoneNumber = message.PhoneNumber;
            IsApproved = message.IsApproved;
            Address = new Address(message.Address);
        }

        public ProviderMessage ToMessage()
        {
            return new ProviderMessage()
            {
                Address = Address.ToMessage(),
                Cvr = Cvr,
                Description = Description,
                CompanyName = CompanyName,
                IsApproved = IsApproved,
                PhoneNumber = PhoneNumber,
                User = base.ToMessage()
            };
        }
        public void AddLink(string id, Link link)
        {
            if (!Links.ContainsKey(id))
            {
                Links.Add(id, link);
            }
        }
        
    }
}

File: Sem03/SEP3/Java/DataServer/src/main/java/com/example/dataserver/models/Provider.java
package com.example.dataserver.models;

import com.google.gson.annotations.SerializedName;
import networking.provider.ProviderMessage;
import org.springframework.lang.Nullable;

import javax.persistence.*;

@Entity
@Table(name = "provider", schema = "sep3")
public class Provider {
    @Id
    @Column(name = "user_id")
    private int id;

    @Nullable
    @Column(name = "company_name")
    private String companyName;

    @Nullable
    @Column(name = "cvr")
    private int cvr;

    @Nullable
    @Column(name = "phone_number")
    private String phoneNumber;

    @Nullable
    @Column(name = "description")
    private String description;

    @Nullable
    @Column(name = "is_approved")
    private boolean isApproved = false;

    @Nullable
    @ManyToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    private Address address;

    @OneToOne
    @MapsId
    @PrimaryKeyJoinColumn(name = "user_id", referencedColumnName = "user_id")
    private User user;

    public Provider() {
    }

    public Provider(int id, String companyName, int cvr, String phoneNumber, String description,
                    boolean isApproved, Address address, String email, String password) {
        this.id = id;
        this.companyName = companyName;
        this.cvr = cvr;
        this.phoneNumber = phoneNumber;
        this.description = description;
        this.isApproved = isApproved;
        this.address = address;
    }

    public Provider(ProviderMessage message) {
        user = new User(message.getUser());
        companyName = message.getCompanyName();
        cvr = message.getCvr();
        phoneNumber = message.getPhoneNumber();
        description = message.getDescription();
        isApproved = message.getIsApproved();
        address = new Address(message.getAddress());
    }

    public ProviderMessage toMessage() {
        return ProviderMessage.newBuilder().setAddress(address.toMessage()).setUser(user.toMessage())
                .setCompanyName(companyName).setCvr(cvr).setPhoneNumber(phoneNumber)
                .setDescription(description).setIsApproved(isApproved).build();
    }


    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getCompanyName() {
        return companyName;
    }

    public void setCompanyName(@Nullable String companyName) {
        this.companyName = companyName;
    }

    public int getCvr() {
        return cvr;
    }

    public void setCvr(int cvr) {
        this.cvr = cvr;
    }

    @Nullable
    public String getPhoneNumber() {
        return phoneNumber;
    }

    public void setPhoneNumber(@Nullable String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    @Nullable
    public String getDescription() {
        return description;
    }

    public void setDescription(@Nullable String description) {
        this.description = description;
    }

    public boolean isApproved() {
        return isApproved;
    }

    public void setApproved(boolean approved) {
        isApproved = approved;
    }

    @Nullable
    public Address getAddress() {
        return address;
    }

    public void setAddress(@Nullable Address address) {
        this.address = address;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    @Override
    public String toString() {
        return "Provider{" + "id=" + id + ", companyName='" + companyName + '\'' + ", cvr=" + cvr
                + ", phoneNumber='" + phoneNumber + '\'' + ", description='" + description + '\''
                + ", isApproved=" + isApproved + ", address=" + address + '}';
    }
}
